Source/CarthageKit/NewResolver.swift

```diff
@@ -0,0 +1,552 @@
+		self.resolvedGitReference = resolvedGitReference
+	}
+
+	/// Attempts to determine the latest valid version to use for each
+	/// dependency in `dependencies`, and all nested dependencies thereof.
+	///
+	/// Sends a dictionary with each dependency and its resolved version.
+	public func resolve(
+		dependencies: [Dependency: VersionSpecifier],
+		lastResolved: [Dependency: PinnedVersion]? = nil,
+		dependenciesToUpdate: [String]? = nil
+		) -> SignalProducer<[Dependency: PinnedVersion], CarthageError> {
+		let result = process(dependencies: dependencies, in: DependencyGraph(whitelist: dependenciesToUpdate, lastResolved: lastResolved))
+			.map { graph -> [Dependency: PinnedVersion] in
+				guard
+					let dependenciesToUpdate = dependenciesToUpdate,
+					let lastResolved = lastResolved,
+					!dependenciesToUpdate.isEmpty else {
+						return graph.versions()
+				}
+
+				var newVersions = graph.versions(for: dependenciesToUpdate)
+				for (dependency, version) in lastResolved {
+					// If we don't have a new version for it, and it's still in the graph, grab the last resolved version
+					guard newVersions[dependency] == nil && graph.node(for: dependency) != nil else { continue }
+					newVersions[dependency] = version
+				}
+
+				return newVersions
+			}
+
+		return SignalProducer(result: result)
+	}
+
+	/// Error cache for used for this instance of the resolver
+	private let errorCache = ErrorCache()
+
+	/// Produces a lazily computed sequence of valid permutations of `dependencies`
+	/// previously seen errors, in the order they should be tried.
+	///
+	/// In other words, each permutation is an array of DependencyNodes the same length as
+	/// `dependencies`. Each array represents one possible permutation of those
+	/// dependencies (chosen from among the versions that actually exist for
+	/// each).
+	private func nodePermutations(for dependencies: [Dependency: VersionSpecifier], in basisGraph: DependencyGraph, withParent parentNode: DependencyNode?) -> Result<NodePermutations, CarthageError> {
+
+		return SignalProducer<(key: Dependency, value: VersionSpecifier), CarthageError>(dependencies)
+			.flatMap(.concat) { (dependency, specifier) -> SignalProducer<[DependencyNode], CarthageError> in
+				let versionProducer: SignalProducer<PinnedVersion, CarthageError>
+				if case let .gitReference(refName) = specifier {
+					versionProducer = self.resolvedGitReference(dependency, refName)
+				} else {
+					versionProducer = self.versionsForDependency(dependency)
+						.filter { specifier.isSatisfied(by: $0) }
+				}
+
+				return versionProducer
+					.map { DependencyNode(dependency: dependency, proposedVersion: $0, versionSpecifier: specifier, parent: parentNode) }
+					.collect()
+					.attemptMap { nodes in
+						guard !nodes.isEmpty else {
+							return .failure(CarthageError.requiredVersionNotFound(dependency, specifier))
+						}
+						return .success(nodes.sorted())
+				}
```

Comment:
There are a bunch of code style violations like the indentation of this `}`. If you install SwiftLint and build, they'll show up as warnings in Xcode.

Reply:
Hm, that's odd, I have swiftlint installed, and it was running before... maybe my xcode got confused about when to run the script. I'll take a look at all my indentation when I do a merge for conflicts.